api/api_request_logger.go

```diff
@@ -0,0 +1,121 @@
+// Copyright (c) 2024 The VeChainThor developers
+
+// Distributed under the GNU Lesser General Public License v3.0 software license, see the accompanying
+// file LICENSE or <https://www.gnu.org/licenses/lgpl-3.0.html>
+
+package api
+
+import (
+	"bytes"
+	"encoding/json"
+	"io"
+	"net/http"
+	"time"
+
+	log "github.com/inconshreveable/log15"
+)
+
+// RequestLogger logs requests to an output
+type RequestLogger struct {
+	enabled      bool
+	writerChan   chan entry
+	stopChan     chan bool
+	outputWriter func(msg string, ctx ...interface{})
+}
+
+// NewThorRequestLogger applies thor business logic
+// when returning an instance of *RequestLogger
+func NewThorRequestLogger(enabledFlag bool, enableLog bool) *RequestLogger {
+	if enabledFlag {
+		return NewRequestLogger(enabledFlag, log.Info)
+	}
+	if enableLog {
+		return NewRequestLogger(enableLog, log.Debug)
+	}
+	return NewRequestLogger(false, nil)
+}
+
+func NewRequestLogger(enabled bool, logger func(msg string, ctx ...interface{})) *RequestLogger {
+	return &RequestLogger{
+		enabled:      enabled,
+		outputWriter: logger,
+		writerChan:   make(chan entry, 100_000),
+		stopChan:     make(chan bool),
+	}
+}
+
+func (l *RequestLogger) Enabled() bool {
+	return l.enabled
+}
+
+// Handle returns an http handler to ensure requests are syphoned into the writer
+func (l *RequestLogger) Handle(handler http.Handler) http.Handler {
+	fn := func(w http.ResponseWriter, r *http.Request) {
+		// Read and log the body (note: this can only be done once)
+		// Ensure you don't disrupt the request body for handlers that need to read it
+		var bodyBytes []byte
+		var err error
+		if r.Body != nil {
+			bodyBytes, err = io.ReadAll(r.Body)
```

Comment:
Close the body

Reply:
Turns out https://pkg.go.dev/net/http#Request
```
        // For server requests, the Request Body is always non-nil
	// but will return EOF immediately when no body is present.
	// The Server will close the request body. The ServeHTTP
	// Handler does not need to.
```

Reply:
Since you have overwritten the body, possibly memory leak will happen if io.ReadAll fails and not closed

Reply:
It should return before the r.Body is written to.